llvm/lib/Target/AMDGPU/SIPeepholeSDWA.cpp

```diff
@@ -164,12 +164,34 @@ class SDWASrcOperand : public SDWAOperand {
   bool Neg;
   bool Sext;
 
-public:
+protected:
   SDWASrcOperand(MachineOperand *TargetOp, MachineOperand *ReplacedOp,
                  SdwaSel SrcSel_ = DWORD, bool Abs_ = false, bool Neg_ = false,
                  bool Sext_ = false)
-      : SDWAOperand(TargetOp, ReplacedOp),
-        SrcSel(SrcSel_), Abs(Abs_), Neg(Neg_), Sext(Sext_) {}
+      : SDWAOperand(TargetOp, ReplacedOp), SrcSel(SrcSel_), Abs(Abs_),
+        Neg(Neg_), Sext(Sext_) {}
+public:
+  /// Create an SDWASrcOperand as an operand for \p MI from the given arguments
+  /// if \p SrcSel_ and the src_sel0 and src_sel1 operands of \p MI are
+  /// compatible.
+  static std::unique_ptr<SDWAOperand>
+  create(const SIInstrInfo *TII, const MachineInstr &MI,
+         MachineOperand *TargetOp, MachineOperand *ReplacedOp,
+         SdwaSel SrcSel_ = DWORD, bool Abs_ = false, bool Neg_ = false,
+         bool Sext_ = false) {
+    if (TII->isSDWA(MI.getOpcode())) {
+      for (auto SelOpName :
+           {AMDGPU::OpName::src0_sel, AMDGPU::OpName::src1_sel}) {
+        const MachineOperand *NamedOp = TII->getNamedOperand(MI, SelOpName);
+        if (NamedOp && !compatibleSelections(
+                           static_cast<SdwaSel>(NamedOp->getImm()), SrcSel_))
+          return std::unique_ptr<SDWAOperand>(nullptr);
+      }
+    }
+
+    return std::unique_ptr<SDWAOperand>(new SDWASrcOperand(
+        TargetOp, ReplacedOp, SrcSel_, Abs_, Neg_, Sext_));
```

Comment:
make_unique?

I don't know why we need dynamic allocation of these. We should probably fix doing this separately

Reply:
`make_unique` does not work here because I made the constructor `protected`. I could change it but I think it is preferable to keep it this way because this should be the only way to create instances of this class.

Reply:
> I don't know why we need dynamic allocation of these. We should probably fix doing this separately

It is necessary because of the use of the abstract base class `SDWAOperand`. The closed class hierarchy could easily be replaced by a tagged union if you consider this beneficial.

Reply:
Resolving this conversation as it is no longer relevant to the current implementation.